File: src/medium/RotateArray.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function (nums, k) {
    let covered = new Array(nums.length).fill(false);

    for (let i = 0; i < nums.length; i++) {
        if (covered[i]) {
            continue;
        }

        let curIndex = i;
        let curValue = nums[i];

        do {
            let j = (curIndex + k) % nums.length;
            let temp = curValue;
            curValue = nums[j];
            nums[j] = temp;
            covered[j] = true;
            curIndex = j;
        } while (curIndex !== i)
    }
};

File: src/medium/RotateArray.java
package medium;

public class RotateArray {
    public void rotate(int[] nums, int k) {
        for (int i = 0; i < gcd(nums.length, k); i++) {
            int index = i;
            int num = nums[index];
            do {
                index = (index + k) % nums.length;
                int temp = nums[index];
                nums[index] = num;
                num = temp;
            } while (index != i);
        }
    }

    public static int gcd(int a, int b) {
        if (b == 0)
            return a;
        return gcd(b, a % b);
    }
}
